extensions/objects/templates/unicode-string-editor.component.spec.ts

```diff
@@ -0,0 +1,202 @@
+// Copyright 2021 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for unicode string editor.
+ */
+
+import { EventEmitter, SimpleChanges } from '@angular/core';
+import { ExternalSaveService } from 'services/external-save.service';
+import { FormsModule } from '@angular/forms';
+import { UnicodeStringEditorComponent } from './unicode-string-editor.component';
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+
+describe('UnicodeStringEditorComponent', () => {
+  let component: UnicodeStringEditorComponent;
+  let fixture: ComponentFixture<UnicodeStringEditorComponent>;
+  let externalSaveService: ExternalSaveService;
+
+  var externalSaveEventEmitter = new EventEmitter();
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      imports: [FormsModule],
+      declarations: [UnicodeStringEditorComponent]
+    }).compileComponents();
+  }));
+
+  beforeEach(angular.mock.module('oppia'));
+  beforeEach(() => {
+    fixture = TestBed.createComponent(UnicodeStringEditorComponent);
+    component = fixture.componentInstance;
+    externalSaveService = TestBed.get(ExternalSaveService);
+    spyOnProperty(externalSaveService, 'onExternalSave').and.returnValue(
+      externalSaveEventEmitter);
+    component.alwaysEditable = false;
+    component.value = 'random value';
+  });
+
+  afterEach(() => {
+    component.ngOnDestroy();
+  });
+
+  it('should initialize component when interaction rule is selected', () => {
+    spyOn(component, 'closeEditor').and.callThrough();
+    spyOn(component.componentSubscriptions, 'add');
+    component.active = true;
+
+    expect(component.active).toBe(true);
+
+    component.ngOnInit();
+
+    // We cannot test what componentSubscriptions.add has been called with since
+    // externalSaveService is a private variable and inside it is an anonymous
```

Comment:
But we can spy on ExternalSaveService.onExternalSave, right? and provide the service instance through `configureTestingModule`, right?

Reply:
No, if we have to spy then we have to write `spyOn(component.externalSaveService, 'onExternalSave');`, but 
 `externalSaveService` is a private variable so it cannot be spied upon.

Reply:
Sorry, I'm asking to spy on the `ExternalSaveService` and not on the `component.externalSaveService`.

i.e, import `ExternalSaveService` and sypOn(ExternalSaveService,  'onExternalSave'), will that work?

Reply:
No, that won't work. We need to have `component.externalSaveService`.

Reply:
`onExternalSave` is a getter, so have you tried spyOnProperty?

Something around:
 `sypOnProperty(ExternalSaveService, 'onExternalSave', 'get')`

Reply:
Done. Thanks for the tip!